# Basta Fazoolin' (8/13/26)

*Codecademy MLE Path — Python classes. Menus → franchises → businesses.*

Times are 24-hour ints so you can compare them (`11` = 11 am, `16` = 4 pm).

## Making the Menus

`Menu` holds a name, an items dict (`name → price`), and a start/end hour. `__repr__` prints when it's available. `calculate_bill` sums prices for a list of item names.

In [1]:
class Menu:
    def __init__(self, name, items, start_time, end_time):
        self.name = name
        self.items = items
        self.start_time = start_time
        self.end_time = end_time

    def _fmt(self, hour):
        suffix = 'am' if hour < 12 else 'pm'
        h = hour if hour <= 12 else hour - 12
        return f'{h} {suffix}'

    def __repr__(self):
        return f'{self.name} menu available from {self._fmt(self.start_time)} to {self._fmt(self.end_time)}'

    def calculate_bill(self, purchased_items):
        total = 0
        for item in purchased_items:
            total += self.items[item]
        return total


brunch = Menu('brunch', {
    'pancakes': 7.50, 'waffles': 9.00, 'burger': 11.00, 'home fries': 4.50,
    'coffee': 1.50, 'espresso': 3.00, 'tea': 1.00, 'mimosa': 10.50, 'orange juice': 3.50
}, 11, 16)

early_bird = Menu('early-bird', {
    'salumeria plate': 8.00, 'salad and breadsticks (serves 2, no refills)': 14.00,
    'pizza with quattro formaggi': 9.00, 'duck ragu': 17.50,
    'mushroom ravioli (vegan)': 13.50, 'coffee': 1.50, 'espresso': 3.00,
}, 15, 18)

dinner = Menu('dinner', {
    'crostini with eggplant caponata': 13.00, 'caesar salad': 16.00,
    'pizza with quattro formaggi': 11.00, 'duck ragu': 19.50,
    'mushroom ravioli (vegan)': 13.50, 'coffee': 2.00, 'espresso': 3.00,
}, 17, 23)

kids = Menu('kids', {
    'chicken nuggets': 6.50, 'fusilli with wild mushrooms': 12.00, 'apple juice': 3.00
}, 11, 21)

print(brunch)
print(brunch.calculate_bill(['pancakes', 'home fries', 'coffee']))
print(early_bird.calculate_bill(['salumeria plate', 'mushroom ravioli (vegan)']))

brunch menu available from 11 am to 4 pm
13.5
21.5


## Creating the Franchises

A `Franchise` is an address plus the menus it serves. `available_menus(time)` returns every menu whose window covers that hour (`start <= time <= end`).

In [2]:
class Franchise:
    def __init__(self, address, menus):
        self.address = address
        self.menus = menus

    def __repr__(self):
        return self.address

    def available_menus(self, time):
        available = []
        for menu in self.menus:
            if menu.start_time <= time <= menu.end_time:
                available.append(menu)
        return available


menus = [brunch, early_bird, dinner, kids]
flagship_store = Franchise('1232 West End Road', menus)
new_installment = Franchise('12 East Mulberry Street', menus)

print(flagship_store)
print(flagship_store.available_menus(12))
print(flagship_store.available_menus(17))

1232 West End Road
[brunch menu available from 11 am to 4 pm, kids menu available from 11 am to 9 pm]
[early-bird menu available from 3 pm to 6 pm, dinner menu available from 5 pm to 11 pm, kids menu available from 11 am to 9 pm]


## Creating Businesses

A `Business` is a name plus a list of franchises. Same `Menu` / `Franchise` classes get reused for Take a' Arepa.

In [3]:
class Business:
    def __init__(self, name, franchises):
        self.name = name
        self.franchises = franchises


basta = Business("Basta Fazoolin' with My Heart", [flagship_store, new_installment])

arepas_menu = Menu("Take a' Arepa", {
    'arepa pabellon': 7.00, 'pernil arepa': 8.50,
    'guayanes arepa': 8.00, 'jamon arepa': 7.50
}, 10, 20)

arepas_place = Franchise('189 Fitzgerald Avenue', [arepas_menu])
arepa = Business("Take a' Arepa", [arepas_place])

print(basta.name, basta.franchises)
print(arepas_menu)
print(arepas_place)
print(arepa.name, arepa.franchises)

Basta Fazoolin' with My Heart [1232 West End Road, 12 East Mulberry Street]
Take a' Arepa menu available from 10 am to 8 pm
189 Fitzgerald Avenue
Take a' Arepa [189 Fitzgerald Avenue]


## TL;DR

- **`Menu`**: name, items dict, 24-hour window. `__repr__` shows hours; `calculate_bill` sums item prices.
- **`Franchise`**: address + menus. `available_menus(time)` filters by `start <= time <= end`.
- **`Business`**: name + list of franchises. Same three classes scale from one restaurant to a second brand (Take a' Arepa).
- Checks: brunch pancakes + home fries + coffee = **13.5**; early-bird salumeria + vegan ravioli = **21.5**; noon → brunch + kids; 5 pm → early-bird + dinner + kids.